### PR TITLE
[Feature]Add tis fall back for ppo_policy_loss

### PR DESCRIPTION
**Pull request by garyzhang99**
## Description

The Truncate Large IS feature addresses training instability issues in PPO caused by computation errors between VLLM (used for `old_logprob`) and transformer (used for `logprob`) calculations.

### Problem Statement

When probabilities are very small, even minor computation errors between VLLM and transformer can lead to extremely large importance sampling ratios. This is particularly problematic when:

1. Both `old_logprob` and `logprob` are small (representing low probabilities)
2. The advantage is negative
3. The ratio becomes very large (e.g., tens or hundreds) due to computation discrepancies

In the standard PPO implementation with one-sided clipping (multiply advantage first, then clip and take max), this can cause:
- Large negative gradients
- Training instability
- Issues like repetition in generated text

### Solution

The Truncate Large IS feature adds a pre-clipping step that truncates the importance sampling ratio to a configurable range before it's used in loss computation. This is similar to CISPO's approach to handling importance sampling ratios.

#### Parameters

- `truncate_large_is` (bool, default: `False`): Enable/disable the truncate large IS feature
- `truncate_is_range_low` (float, default: `0.0`): Lower bound for IS ratio truncation
- `truncate_is_range_high` (float, default: `2.0`): Upper bound for IS ratio truncation

#### Usage in YAML Config

```yaml
algorithm:
  algorithm_type: grpo
  ...
  policy_loss_fn_config:
    clip_range: 0.2
    loss_agg_mode: token-mean
    # Enable truncate large IS
    truncate_large_is: true
    truncate_is_range_low: 0.0
    truncate_is_range_high: 2.0
```

#### Recommended Settings

For general use with potential VLLM/transformer computation errors:
```yaml
truncate_large_is: true
truncate_is_range_low: 0.0
truncate_is_range_high: 2.0
```

### Implementation Details

The truncation is applied as follows:

```python
# Compute ratio
ratio = torch.exp(logprob - old_logprob)

# Apply truncation if enabled
if truncate_large_is:
    ratio = torch.clamp(ratio, truncate_is_range_low, truncate_is_range_high)

# Continue with standard PPO loss computation
pg_losses = -advantages * ratio
pg_losses2 = -advantages * torch.clamp(ratio, 1.0 - clip_range_low, 1.0 + clip_range_high)
pg_loss = torch.max(pg_losses, pg_losses2)
```

This ensures that extremely large (or small) ratios are bounded before they interact with the advantage values.

### When to Use

Consider enabling this feature when:
1. Training exhibits instability (loss spikes, large grad norms)
2. Generated text shows repetition or degeneration
3. You observe extremely large IS ratios in training

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 1]
```diff
@@ -20,7 +20,24 @@ def __init__(
...
+        Initialize PPO policy loss function.
+
+        Args:
+            backend: Backend framework (default: "verl")
+            clip_range: Symmetric clipping range for PPO
+            clip_range_low: Lower bound for clipping (1.0 - clip_range_low)
+            clip_range_high: Upper bound for clipping (1.0 + clip_range_high)
+            loss_agg_mode: Loss aggregation mode (default: "token-mean")
+            truncate_large_is: Whether to truncate large importance sampling ratios
+                to handle computation errors between VLLM and transformer calculations
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]:
```suggestion
                to handle calculation discrepancies between rollout and training engines
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 2]
```diff
@@ -20,7 +20,24 @@ def __init__(
         clip_range_low: Optional[float] = None,
         clip_range_high: Optional[float] = None,
         loss_agg_mode: Optional[str] = "token-mean",
+        truncate_large_is: bool = False,
+        truncate_is_range_low: Optional[float] = None,
+        truncate_is_range_high: Optional[float] = None,
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]:
```suggestion
        truncate_is_range_low: float = 0.0,
        truncate_is_range_high: float = 2.0,
```

Comment (yanxi-chen) [reply trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: Would it be better to set default values here? (also more consistent with current implementation of other policy loss functions)

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 3]
```diff
@@ -34,6 +51,23 @@ def __init__(
...
+            self.truncate_is_range_high = (
+                truncate_is_range_high if truncate_is_range_high is not None else 2.0
+            )
+            assert self.truncate_is_range_low >= 0.0, "truncate_is_range_low must be non-negative."
+            assert (
+                self.truncate_is_range_high > self.truncate_is_range_low
+            ), "truncate_is_range_high must be greater than truncate_is_range_low."
+        else:
+            self.truncate_is_range_low = None
+            self.truncate_is_range_high = None
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: These lines seem unnecessary

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 4]
```diff
@@ -34,6 +51,23 @@ def __init__(
...
 
+        # Truncate large IS configuration
+        self.truncate_large_is = truncate_large_is
+        if truncate_large_is:
+            self.truncate_is_range_low = (
+                truncate_is_range_low if truncate_is_range_low is not None else 0.0
+            )
+            self.truncate_is_range_high = (
+                truncate_is_range_high if truncate_is_range_high is not None else 2.0
+            )
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: Remove 0.0 and 2.0 if these default values have been set in input to init method

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 5]
```diff
@@ -46,6 +80,18 @@ def __call__(  # type: ignore
         ratio = torch.exp(negative_approx_kl)
         ppo_kl = masked_mean(-negative_approx_kl, action_mask)
 
+        # Truncate large IS ratios if enabled
+        # This helps stabilize training when there are computation errors between
+        # VLLM and transformer calculations, especially for small probabilities
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]:
```suggestion
        # This helps stabilize training when there are computation discrepancies between
        # rollout and training engines, especially for small probabilities
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 6]
```diff
@@ -46,6 +80,18 @@ def __call__(  # type: ignore
...
+        if self.truncate_large_is:
+            ratio_before_truncate = ratio.clone()
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: Cloning `ratio` merely for calculating `is_truncate_frac` seems a bit too heavy. How about something like `is_truncate_frac = masked_mean(ratio < self.truncate_is_range_low) +  masked_mean(ratio > self.truncate_is_range_high)`? (might need to handle detach appropriately)

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py (hunk 2 above)
Comment (garyzhang99) [reply trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: We set default values here.

**Comment by hiyuchang** [comment]
Solved by #388
